modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EUnary.java

```diff
@@ -68,90 +57,4 @@ public Operation getOperation() {
     public <Scope> void visitChildren(UserTreeVisitor<Scope> userTreeVisitor, Scope scope) {
         childNode.visit(userTreeVisitor, scope);
     }
-
-    @Override
-    void analyze(SemanticScope semanticScope) {
-        if (semanticScope.getCondition(this, Write.class)) {
-            throw createError(new IllegalArgumentException(
-                    "invalid assignment: cannot assign a value to " + operation.name + " operation " + "[" + operation.symbol + "]"));
-        }
-
-        if (semanticScope.getCondition(this, Read.class) == false) {
-            throw createError(new IllegalArgumentException(
-                    "not a statement: result not used from " + operation.name + " operation " + "[" + operation.symbol + "]"));
-        }
-
-        Class<?> valueType;
-        Class<?> promote = null;
-
-        if ((operation == Operation.SUB || operation == Operation.ADD) &&
-                (childNode instanceof ENumeric || childNode instanceof EDecimal)) {
-            semanticScope.setCondition(childNode, Read.class);
-            semanticScope.copyDecoration(this, childNode, TargetType.class);
-            semanticScope.replicateCondition(this, childNode, Explicit.class);
-            semanticScope.replicateCondition(this, childNode, Internal.class);
-
-            if (childNode instanceof ENumeric) {
-                ENumeric numeric = (ENumeric)childNode;
-
-                if (operation == Operation.SUB) {
-                    numeric.analyze(semanticScope, numeric.getNumeric().charAt(0) != '-');
-                } else {
-                    childNode.analyze(semanticScope);
-                }
-            } else if (childNode instanceof EDecimal) {
-                EDecimal decimal = (EDecimal)childNode;
-
-                if (operation == Operation.SUB) {
-                    decimal.analyze(semanticScope, decimal.getDecimal().charAt(0) != '-');
-                } else {
-                    childNode.analyze(semanticScope);
-                }
-            } else {
-                throw createError(new IllegalArgumentException("illegal tree structure"));
-            }
-
-            valueType = semanticScope.getDecoration(childNode, ValueType.class).getValueType();
```

Comment:
Why is this different in DefaultSemanticAnalysisPhase?  Is it just the changes to use decorations?

Reply:
Yes, this what changed by adding the Negate decoration. Parsing a negative numeric constant (and decimal constant) is a bit odd as the parser gives you the tokens MINUS and NUMERIC, but if they're separated a user cannot enter the negative maximum value as the positive one exceeds the positive max since the negation won't be applied until a parent node. Special care is taken to ensure this doesn't happen that way by collapsing the unary node with the Negate decoration.